File: backend/src/main/java/com/backend/ord/domain/entities/Word.java
package com.backend.ord.domain.entities;

import com.backend.ord.domain.embedded.ExampleSentence;
import com.backend.ord.domain.entities.abstracts.EntityBase;
import com.backend.ord.enums.Language.LanguageName;
import com.backend.ord.enums.Word.WordType;
import jakarta.persistence.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.hibernate.type.SqlTypes;

import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "words")
public class Word extends EntityBase {
    @Column(name = "origin", nullable = false)
    private String origin;

    @Column(name = "translation", nullable = false)
    private String translation;

    @Column(name = "is_bookmarked", nullable = false)
    private Boolean isBookmarked = false;

    @Column(name = "points", nullable = false)
    private Integer points = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "bank_id")
    private Bank bank;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "type", nullable = false)
    @Enumerated(EnumType.STRING)
    private WordType type;

    @Column(name = "translated_from", nullable = false)
    @Enumerated(EnumType.STRING)
    private LanguageName translatedFrom;

    @Column(name = "translated_to", columnDefinition = "language_name(0, 0) not null")
    @Enumerated(EnumType.STRING)
    private LanguageName translatedTo;

    @Column(name = "example_sentences", nullable = false)
    @JdbcTypeCode(SqlTypes.JSON)
    private Set<ExampleSentence> exampleSentences = new HashSet<ExampleSentence>();

    public String getOrigin() {
        return this.origin;
    }

    public String getTranslation() {
        return this.translation;
    }

    public Boolean getIsBookmarked() {
        return this.isBookmarked;
    }

    public Integer getPoints() {
        return this.points;
    }

    public Bank getBank() {
        return this.bank;
    }

    public User getUser() {
        return this.user;
    }

    public WordType getType() {
        return this.type;
    }

    public LanguageName getTranslatedFrom() {
        return this.translatedFrom;
    }

    public LanguageName getTranslatedTo() {
        return this.translatedTo;
    }

    public Set<ExampleSentence> getExampleSentences() {
        return this.exampleSentences;
    }

    public void setOrigin(String origin) {
        this.origin = origin;
    }

    public void setTranslation(String translation) {
        this.translation = translation;
    }

    public void setIsBookmarked(Boolean isBookmarked) {
        this.isBookmarked = isBookmarked;
    }

    public void setPoints(Integer points) {
        this.points = points;
    }

    public void setBank(Bank bank) {
        this.bank = bank;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public void setType(WordType type) {
        this.type = type;
    }

    public void setTranslatedFrom(LanguageName translatedFrom) {
        this.translatedFrom = translatedFrom;
    }

    public void setTranslatedTo(LanguageName translatedTo) {
        this.translatedTo = translatedTo;
    }

    public void setExampleSentences(Set<ExampleSentence> exampleSentences) {
        this.exampleSentences = exampleSentences;
    }
}

File: backend/src/main/kotlin/com/backend/ord/domain/entities/Word.kt
package com.backend.ord.domain.entities

import com.backend.ord.domain.embedded.ExampleSentence
import com.backend.ord.domain.entities.interfaces.IdentifiableUserResource
import com.backend.ord.enums.Language.LanguageName
import com.backend.ord.enums.Word.WordExtraMark
import com.backend.ord.enums.Word.WordType
import jakarta.persistence.*
import jakarta.persistence.Table
import org.hibernate.annotations.*
import org.hibernate.type.SqlTypes
import java.time.Instant
import java.util.*

@Entity
@Table(name = "words")
class Word(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    override var id: UUID = UUID.randomUUID(),

    @Column(name = "origin", nullable = false)
    var origin: String,

    @Column(name = "translation", nullable = false)
    var translation: String,

    @Column(name = "definition", nullable = false)
    var definition: String,

    @Column(name = "use_cases", nullable = false)
    @JdbcTypeCode(SqlTypes.JSON)
    var useCases: Set<String> = emptySet(),

    @Column(name = "is_bookmarked", nullable = false)
    var isBookmarked: Boolean = false,

    @Column(name = "points", nullable = false)
    var points: Int = 0,

    @Column(name = "type", nullable = false)
    @Enumerated(EnumType.STRING)
    var type: WordType,

    @Column(name = "extra_mark", nullable = false)
    @Enumerated(EnumType.STRING)
    var extraMark: WordExtraMark? = null,

    @Column(name = "translated_from", nullable = false)
    @Enumerated(EnumType.STRING)
    var translatedFrom: LanguageName,

    @Column(name = "translated_to", columnDefinition = "language_name(0, 0) not null")
    @Enumerated(EnumType.STRING)
    var translatedTo: LanguageName,

    @Column(name = "example_sentences", nullable = false)
    @JdbcTypeCode(SqlTypes.JSON)
    var exampleSentences: Set<ExampleSentence> = emptySet(),

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    override var user: User,

    @Column(name = "user_id", insertable = false, updatable = false)
    var userId: UUID = user.id,

    @ManyToOne(fetch = FetchType.LAZY, optional = true)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "bank_id", nullable = true)
    var bank: Bank? = null,

    @Column(name = "bank_id", nullable = true, insertable = false, updatable = false)
    var bankId: UUID? = bank?.id,

    @Column(name = "bank_group_id", nullable = true, updatable = false)
    var bankGroupId: UUID? = bank?.bankGroupId,

    @Column(name = "created_at", nullable = false, updatable = false)
    @CreationTimestamp
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false, updatable = false)
    @UpdateTimestamp
    var updatedAt: Instant = Instant.now()
) : IdentifiableUserResource {
    @PostLoad
    fun populateUserId() {
        userId = user.id
        bankId = bank?.id
        bankGroupId = bank?.bankGroupId
    }
}

